### PR TITLE
Hook up std::net to wasi-libc on wasm32-wasip2 target

### PR DESCRIPTION
**Pull request by nickrum**
One of the improvements of the `wasm32-wasip2` target over `wasm32-wasip1` is better support for networking. Right now, p2 is just re-using the `std::net` implementation from p1. This PR adds a new net module for p2 that makes use of net from `sys_common` and calls wasi-libc functions directly.

There are currently a few limitations:

- Duplicating a socket is not supported by WASIp2 (directly returns an error)
- Peeking is not yet implemented in wasi-libc (we could let wasi-libc handle this, but I opted to directly return an error instead)
- Vectored reads/writes are not supported by WASIp2 (the necessary functions are available in wasi-libc, but they call WASIp1 functions which do not support sockets, so I opted to directly return an error instead)
- Getting/setting `TCP_NODELAY` is faked in wasi-libc (uses the fake implementation instead of returning an error)
- Getting/setting `SO_LINGER` is not supported by WASIp2 (directly returns an error)
- Setting `SO_REUSEADDR` is faked in wasi-libc (since this is done from `sys_common`, the fake implementation is used instead of returning an error)
- Getting/setting `IPV6_V6ONLY` is not supported by WASIp2 and will always be set for IPv6 sockets (since this is done from `sys_common`, wasi-libc will return an error)
- UDP broadcast/multicast is not supported by WASIp2 (since this is configured from `sys_common`, wasi-libc will return appropriate errors)
- The `MSG_NOSIGNAL` send flag is a no-op because there are no signals in WASIp2 (since explicitly setting this flag would require a change to `sys_common` and the result would be exactly the same, I opted to not set it)

Do those decisions make sense?

While working on this PR, I noticed that there is a `std::os::wasi::net::TcpListenerExt` trait that adds a `sock_accept()` method to `std::net::TcpListener`. Now that WASIp2 supports standard accept, would it make sense to remove this?

cc @alexcrichton

### CODE REVIEW

**Comment by tgross35** [comment]
Thanks for the PR! The reasoning for these changes makes sense, but why are all the extern interfaces added here? These changes should go into our libc https://github.com/rust-lang/libc/blob/72cb7aaf50bf70b5d360b1a64b0d52d3ed3fbf15/src/wasi.rs.

Pinging was[mi] expert @alexcrichton, could you double check the reasoning here?

**Comment by tgross35** [comment]
More specifically, if these are available on wasi-p2 but not p1 then the libc changes would mean:

- Move `wasi.rs` to `wasi/mod.rs`
- Create `wasi/p2.rs` that includes things only available on `p2`
- If there is anything that is available on `p1` but not `p2`, move it to `wasi/p1.rs`
- In the new `wasi/mod.rs`, import everything from either `p1` or `p2` based on the target (like https://github.com/rust-lang/libc/blob/72cb7aaf50bf70b5d360b1a64b0d52d3ed3fbf15/src/unix/mod.rs#L1570-L1612)

**Comment by nickrum** [comment]
I was planning to work on bringing those extern interfaces to [rust-lang/libc](https://github.com/rust-lang/libc) next by doing exactly that.
Since splitting libc between WASIp1 and WASIp2 and verifying which types/structs/functions are available where felt like a bigger effort, since there already is a `netc` module inside WASIp1's net and since there are other targets having their network-related extern interfaces directly within std, my thought was that it might be fine to add them here for the time being to get things started and replace them once the changes to libc are ready. But if you prefer to have them in libc first, that makes total sense as well.

**Comment by tgross35** [comment]
It would be better to update libc first so definitions only live in a single place, I would be concerned about things getting out of sync otherwise. Doing that shouldn't be much worse than adding them here - you can probably just do steps 1, 2, and 4 in my list above, skipping step 3 unless it turns out to be a problem.

The turnaround is pretty quick getting a new libc version and updating std.

**Comment by tgross35** [comment]
Also if you have a link to the headers please post them in the libc PR, sometimes it's tough to track down where definitions came from

**Comment by alexcrichton** [comment]
r? @alexcrichton 

Thanks so much for working on this! I'll do a review tomorrow

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:71]: The WASIp2 target is moving in a different direction than WASIp1 so one refactoring I've been meaning to do for awhile is to reduce the reliance on `WasiFd` in the WASIp2 target. To that effect could you switch this to directly using a `c_int` internally? That would require reproduction of the destruction and removal of the `AsRef<WasiFd>` implementation but that accurately reflects how a socket works on this platform.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:188]: You mentioned that this is implemented in wasi-libc but returns an error there, could this be forwarded to wasi-libc to have wasi-libc return the error?

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:196]: For this instead of `read_vectored` this'll want to do the default implementation of finding the first non-empty buffer and using that for a read. That's still a correct implementation for this function just not as "vectorized"

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:230]: Similar to above could this call libc functions to have errors returned through there?

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:234]: Similar to above, could this find the first non-empty buffer and write that?

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:252]: Could this perhaps be: `let secs = dur.as_secs().try_into().unwrap_or(netc::time_t::MAX)`

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:259]: I realize that this was probably copy/pasted from elsewhere and adapted for wasi but is this needed? I would have thought this would be caught by the first condition above

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [inline library/std/src/sys/pal/wasip2/net.rs:294]: Is it possible to call set/getsockopt and let libc return the error here?

**Review by alexcrichton — COMMENTED** [review]
Thank you so much for working on this! I've left some comments below with a few thoughts here and there. Mostly I'm hoping we can delegate to libc to return errors where possible instead of explicitly returning errors here since that'll make it easier to support in the future by just changing wasi-libc. Otherwise most of your other changes also makes sense to me, but let me know if you have any more questions.

As for `libc` bindings, I agree it'd be great to get these bindings upstream into the `libc` crate itself. I can try to take a look at setting that up CI-wise, and/or please feel free to cc me on PRs.

**Comment by alexcrichton** [comment]
I've started a bit with wasi-libc by submitting https://github.com/rust-lang/libc/pull/3869 to re-enable CI. Once that's ready I'll also submit a new target of wasm32-wasip2 and get that configured/ready too. After that it should be a simpler to add p2-specific constants and such.

**Comment by alexcrichton** [comment]
Oh I can also mention, have you been able to test this? For example are you able to run perhaps the `std::net`-related portions of the testsuite in the standard library? The p2 target isn't tested on CI so it's just for local testing and double-checking, but one day the p1 tests may migrate to p2 tests in CI.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (nickrum) [reply library/std/src/sys/pal/wasip2/net.rs]: This is actually how I had it before. I changed it to directly return `unsupported()` to make it more explicit, that this is not (yet) supported. Making it easier to support things in the future by just changing wasi-libc sounds like a very good idea, so I changed it back to let wasi-libc return the error 👍

File: library/std/src/sys/pal/wasip2/net.rs
Comment (nickrum) [reply library/std/src/sys/pal/wasip2/net.rs]: Unfortunately, I had to duplicate the `write()` method, because it is implemented on `TcpStream` in `sys_common` instead of on `Socket`, but this should work now.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (nickrum) [reply library/std/src/sys/pal/wasip2/net.rs]: I indeed adapted most methods from the implementation for UNIX or other targets. If I understand it correctly, this condition effectively rounds up any timeout that is lower than 1 microsecond but greater or equal to 1 nanosecond to 1 microsecond.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (nickrum) [reply library/std/src/sys/pal/wasip2/net.rs:300]: Unfortunately, `SO_LINGER` is not (yet) defined in wasi-libc. Would it make sense to add the constant to wasi-libc before this PR is merged or do we want to keep it like this for the time being?

**Comment by tgross35** [comment]
> Thanks for the review and starting the work on libc. I'll have a look at moving over the brindings next.
> 
> Unfortunately, building the standard library test suit for the `wasm32-wasip2` target fails for me, so I was only able to test this outside of the std lib. Is there a way to only build the std::net-related tests? Otherwise I can look into the failing tests as well.

You should be able to filter them with something like `./x t library/std -- net`. But do they fail to build or fail to run? I'm not sure why there would be a build failure, but if you post some logs we might be able to help.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [reply library/std/src/sys/pal/wasip2/net.rs:71]: Hm ok if it's difficult to do then it's fine to leave this in for now. The WASIp2 traits in this regard need to be improved and is something I should put on my plate of things-to-refactor for the future. My guess is that it's probably too much to take on in this PR, so feel free to defer it.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [reply library/std/src/sys/pal/wasip2/net.rs:300]: Ah ok, yeah if `SO_LINGER` isn't defined at all then it's fine to omit this and leave it as `unsupported()` for now.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [reply library/std/src/sys/pal/wasip2/net.rs:265]: Ah I see, makes sense!

**Comment by alexcrichton** [comment]
Ok all looks good to me, thanks again! Before landing though I think it'd be good to run this through its paces with the test suite of the standard library (or another library if you're interested in another library perhaps?). If the standard library's tests don't build at all is it possible to temporarily comment out the pieces that don't build? This won't be run on CI anyway so it's more of a one-time check to make sure things are working.

**Comment by tgross35** [comment]
Are the bindings able to go into libc? I'm working on rolling a release today

**Comment by nickrum** [comment]
I'm working on the bindings right now. If you could hold off on the release, that'd be great.

**Comment by tgross35** [comment]
No problem! They're pretty quick to do whenever

**Comment by nickrum** [comment]
> But do they fail to build or fail to run? I'm not sure why there would be a build failure, but if you post some logs we might be able to help.

They fail to build. Here are the errors I'm getting:

```rust
error: cannot find macro `error` in this scope
    --> std/src/fs/tests.rs:1147:5
     |
1147 |     error!(blank.create(true).open(&tmpdir.join("f")), invalid_options);
     |     ^^^^^

error[E0433]: failed to resolve: could not find `unix` in `os`
   --> std/src/process/tests.rs:574:16
    |
574 |     crate::os::unix::process::CommandExt::arg0(&mut command, "exciting-name");
    |                ^^^^ could not find `unix` in `os`
    |
note: found an item that was configured out
   --> std/src/os/mod.rs:27:9
    |
27  | pub mod unix {}
    |         ^^^^
note: the item is gated here
   --> std/src/os/mod.rs:19:1
    |
19  | / #[cfg(all(
20  | |     doc,
21  | |     any(
22  | |         all(target_arch = "wasm32", not(target_os = "wasi")),
23  | |         all(target_vendor = "fortanix", target_env = "sgx")
24  | |     )
25  | | ))]
    | |___^
note: found an item that was configured out
   --> std/src/os/mod.rs:65:9
    |
65  | pub mod unix;
    |         ^^^^
note: the item is gated here
   --> std/src/os/mod.rs:64:1
    |
64  | #[cfg(all(not(target_os = "hermit"), any(unix, doc)))]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0405]: cannot find trait `AsRawFd` in this scope
   --> std/src/net/tcp/tests.rs:678:32
    |
678 |     fn render_inner(addr: &dyn AsRawFd) -> impl fmt::Debug {
    |                                ^^^^^^^ not found in this scope
    |
help: consider importing one of these traits
    |
1   + use crate::os::fd::AsRawFd;
    |
1   + use crate::realstd::os::fd::AsRawFd;
    |

error[E0425]: cannot find function `symlink_file` in this scope
  --> std/src/fs/tests.rs:76:11
   |
76 |     match symlink_file(r"nonexisting_target", link) {
   |           ^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `junction_point` in this scope
   --> std/src/fs/tests.rs:595:12
    |
595 |     check!(junction_point(&d2, &dt.join("d2")));
    |            ^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `symlink_dir` in this scope
   --> std/src/fs/tests.rs:702:17
    |
702 |         let _ = symlink_dir(attack_dest_dir, &victim_del_path);
    |                 ^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `env_cmd` in this scope
   --> std/src/process/tests.rs:276:19
    |
276 |     let mut cmd = env_cmd();
    |                   ^^^^^^^ not found in this scope
```

It looks like the standard library test suite expects the `target_family` to be either `unix` or `windows`, which is not true for the `wasm` targets.

> Before landing though I think it'd be good to run this through its paces with the test suite of the standard library (or another library if you're interested in another library perhaps?). If the standard library's tests don't build at all is it possible to temporarily comment out the pieces that don't build?

I agree, that would be good! I can't think of a library to thoroughly test this with off the top of my head, but I'll look into getting the test suite to build 👍

**Comment by alexcrichton** [comment]
Ah alas I forgot that many of the sockets-related tests all used threads as well. Were many able to be run under wasm32-wasip2? If not that's ok, but thanks for getting the tests compiling at least!

**Comment by nickrum** [comment]
> Ah ok, thanks for the explanation! In that case I think it's ok to comment things out locally enough to get the `std::net` tests to run for wasi, and then in the future there can perhaps be work-in-parallel to upstream changes to the standard library's tests to prepare to run on non-unix/non-windows platforms as well to make it easier in the future?

Getting the tests to build turned out to be easier than I initially thought since those tests that expect either a `unix` or a `windows` environment are all related to the file system and I had to ignore them anyway because they rely on `std::env::temp_dir()`, which just panics on WASI targets.

> Ah alas I forgot that many of the sockets-related tests all used threads as well. Were many able to be run under wasm32-wasip2? If not that's ok, but thanks for getting the tests compiling at least!

Yeah, there aren't many sockets-related tests left, unfortunately. But those that do not rely on threads are all running successfully (at least for me locally).

With the last push, I also disabled network-related tests for WASIp1 and changed the default config for the wasi test runner to enable networking and allow IP name lookup.

File: library/std/src/sys/pal/wasip2/net.rs
Comment (alexcrichton) [reply library/std/src/sys/pal/wasip2/net.rs:71]: If that's easy enough yeah `OwnedFd` I think would be a better fit since we need to eventually sunset `WasiFd` on the p2 target anyway. Again though feel free to defer to a future PR. The main thing I was worried about which was `std::os::wasi` is already unstable on the p2 target, as desired, so what I was originally worried about is already taken care of.

**Review by alexcrichton — APPROVED** [review]
This all looks great to me, thanks again for doing the work on getting the libstd tests working! 

@tgross35 do you have an opinion one way or another on landing this before or after a libc update in the workspace here? I don't have a strong feeling myself but if it's desired to not duplicate these libc definitions then I think the only remaining blocker is waiting on the libc update

**Comment by jeffparsons** [comment]
Just checking whether my understanding of the state of this PR is correct. I believe what's needed is:

- [ ] A libc 0.2.x release containing https://github.com/rust-lang/libc/pull/3899, which has already been merged into the `libc-0.2` branch.
- [ ] Bump libc version in this repo (a la https://github.com/rust-lang/rust/pull/129538) to that new libc release.

Does that sound right?

**Comment by nickrum** [comment]
@jeffparsons I'm not exactly sure if we have to wait for a weekly `cargo update` PR or if I could just update `libc` in this PR once a new release is there. Other than that, this sounds right.

**Comment by jeffparsons** [comment]
@tgross35 What's involved in getting a new libc release published? Is it just done periodically, done on request, or...?

(Sorry if this is naggy — it just wasn't clear to me if the default for this sort of thing is meant to be "be patient" or "things only happen if asked". I am enthusiastically open to feedback on this! 🙇‍♂️)

**Comment by tgross35** [comment]
Sorry about that, it is indeed waiting on me :) I was waiting on a fix to get in that took a little while, now I just need to double check everything and actually do the release but haven't had time (I'll be back early next week, will be easier then).

> (Sorry if this is naggy — it just wasn't clear to me if the default for this sort of thing is meant to be "be patient" or "things only happen if asked". I am enthusiastically open to feedback on this! 🙇‍♂️)

No worries! I've been trying to do one every few PRs or whenever there is a logical group of changes. This one has been more or less at that point for a couple weeks, just got delayed by needing a fix.

**Comment by tgross35** [comment]
Done :) https://github.com/rust-lang/libc/releases/tag/0.2.159

The libc bump should be in its own PR, feel free to `r?` me if anyone puts one up

**Comment by jeffparsons** [comment]
@nickrum It looks like the required libc bump was included in https://github.com/rust-lang/rust/pull/130892 so I think this just needs a rebase! 😀

File: library/std/src/sys/pal/wasip2/net.rs
Comment (nickrum) [reply library/std/src/sys/pal/wasip2/net.rs:74]: Alright, let's defer it then to not block this PR any further.

**Comment by nickrum** [comment]
Should be good to go now 👍

**Pull request merged**
